File: CodingInterviewsJavaSolutions/Linked List Cycle II.java
public class Solution {
    public ListNode transection(ListNode node) {
        ListNode tortoise = node;
        ListNode hare = node;
        while (hare != null && hare.next != null) {
            tortoise = tortoise.next;
            hare = hare.next.next;
            if (tortoise == hare) {
                return tortoise;
            }
        }
        return null;
    }

    public ListNode detectCycle(ListNode head) {
        if (head == null) {
            return head;
        }

        ListNode trans = transection(head);
        if (trans == null) {
            return trans;
        }

        while (trans != head) {
            head = head.next;
            trans = trans.next;
        }
        return trans;
    }
}

File: CodingInterviewsPythonSolutions/Linked List Cycle II.py
class Solution:
    def transection(self, node):
        tortoise = node
        hare = node
        while hare and hare.next:
            tortoise = tortoise.next
            hare = hare.next.next
            if tortoise == hare:
                return tortoise
        return None

    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        trans = self.transection(head)
        if not trans:
            return trans

        while trans != head:
            head = head.next
            trans = trans.next
        return trans
